<h1>這次目標：</h1>

這次嘗試把104人力銀行行銷企劃相關的工作2019的資料拿起來做整理：

1. 想要找出：
	(1)資料總筆數
	(2)行銷企劃需要會使用的工具，主要有哪些
	(3)需要至少“英文中等”筆數

2. 整理出一份新的csv檔案
	欄位名稱如下：
	職稱、公司名稱、縣市、鄉鎮、路名、工作時間、最高需求人數(數字)、最高月薪(數字)、更新年份、更新月份、更新日、要求年資(數字)、
    最低學歷要求

首先引入pandas函式庫，先確認是否成功匯入資料：

In [1]:
import pandas as pd

In [2]:
data = pd.read_csv('104行銷企劃拷貝.csv',sep=',')
data.head(3)

Unnamed: 0 title          cName  \
0           0  行銷企劃      華友旅行社有限公司   
1           1  行銷企劃  羊舍鮮羊乳_健豪農產品商行   
2           2  行銷企劃     吉品養生股份有限公司   

                                             jobCont               jobLoc  \
0  1.負責行銷企劃案的撰寫、規劃，籌備、執行公司對外的行銷活動與媒體活動（如：展覽、促銷活動）...    台北市大安區敦化南路二段76號8樓   
1  1. 品牌行銷策略的擬定與有效執行 2. 官網管理：產品上架、產品文案撰寫與Banner製作...        彰化縣彰化市河濱路475號   
2  我們要尋找台灣飲食產業的開創者   對於飲食有高度的熱情，同時熟知市場動態  關注各種新創商...  台北市大安區忠孝東路4段212號10樓   

                 category vacation           workT hireNum empType  ...  \
0    行銷企劃人員、產品行銷人員、網站行銷企劃     週休二日  日班，09:00~18:00     1 人      全職  ...   
1    行銷企劃主管、行銷企劃人員、產品行銷人員     週休二日              日班     1 人      全職  ...   
2  行銷企劃人員、行銷企劃助理、產品企劃開發人員     週休二日              日班      不限      全職  ...   

                                               tools  \
0  Windows 2000、Excel、Outlook、PowerPoint、Word、中文打...   
1  Excel、PowerPoint、Word、Google Analytics、Adobe P...   
2                                                 不拘   

                                              skills  \
0                                                 不拘   
1  文書處理╱排版能力、行銷製作物規劃與執行、品牌行銷管理、異業合作規劃與執行、提案與簡報技巧、...   
2                                                 不拘   

                                            otherReq  \
0                                                未填寫   
1  【加分條件】 －熟悉SHOPLINE後臺操作 －有拍攝影片及剪接經驗 －善ＧＡ分析 －團購/...   
2  ● 工作需求  1. 產品與專案開發、行銷、策略規劃與分析…等相關工作經驗兩年以上  2. ...   

                                              sabisu insurance entertain  \
0  ◆ 保險類 　1.勞保 　2.健保 ◆ 獎金/禮品類  1.年終獎金  2.三節獎金/禮品 ...       NaN       NaN   
1                                               尚未提供       NaN       NaN   
2  ◆ 保險類  1.勞保  2.健保 ◆ 其他  1.員工購物優惠         2.節慶獎...       NaN       NaN   

  bonus contact                 email similar  
0   NaN     李小姐                   NaN     NaN  
1   NaN     曾小姐  goatbarnmk@gmail.com     NaN  
2   NaN     陳先生                   NaN     NaN  

[3 rows x 29 columns]

先確認資料細節, 有沒有null項和總資料數：

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3010 entries, 0 to 3009
Data columns (total 29 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  3010 non-null   int64  
 1   title       3010 non-null   object 
 2   cName       3010 non-null   object 
 3   jobCont     3010 non-null   object 
 4   jobLoc      3010 non-null   object 
 5   category    3010 non-null   object 
 6   vacation    3010 non-null   object 
 7   workT       3010 non-null   object 
 8   hireNum     3010 non-null   object 
 9   empType     3010 non-null   object 
 10  manage      3010 non-null   object 
 11  busTrip     3010 non-null   object 
 12  salary      3010 non-null   object 
 13  job_update  3010 non-null   object 
 14  accept      3010 non-null   object 
 15  workExp     3010 non-null   object 
 16  edu         3010 non-null   object 
 17  department  3010 non-null   object 
 18  lan         3010 non-null   object 
 19  tools       3010 non-null  

所以可以看到這份資料的總筆數

<h3>總共是：3010筆</h3>

<h2>接下來整理技能tools欄位來看行銷企劃最主要的工作技能有哪些</h2>

首先把每一個工作職缺的技能一個個拆開：

In [4]:
new = data['tools'].str.split("、", expand = True)
new.head(10)

0           1        2                 3                4  \
0  Windows 2000       Excel  Outlook        PowerPoint             Word   
1         Excel  PowerPoint     Word  Google Analytics  Adobe Photoshop   
2            不拘        None     None              None             None   
3            不拘        None     None              None             None   
4            不拘        None     None              None             None   
5            不拘        None     None              None             None   
6         Excel  PowerPoint     Word         中文打字50~75        英文打字50~75   
7            不拘        None     None              None             None   
8            不拘        None     None              None             None   
9            不拘        None     None              None             None   

             5          6     7     8  
0    中文打字20~50  英文打字20~50  None  None  
1  Illustrator       None  None  None  
2         None       None  None  None  
3         None       None  None  None  
4         None       None  None  None  
5         None       None  None  None  
6         None       None  None  None  
7         None       None  None  None  
8         None       None  None  None  
9         None       None  None  None

但每一欄需要被重新整合才能方便統計每個技能的數量，所以重新把打散的欄位合成一個大的list：

In [5]:
alltools = []
for i in range(0,9):
    for j in range(0,3010):
        alltools.append(new[i][j])
print(alltools)

['Windows 2000', 'Excel', '不拘', '不拘', '不拘', '不拘', 'Excel', '不拘', '不拘', '不拘', '不拘', 'Excel', '不拘', '不拘', '不拘', 'PowerPoint', 'Excel', '不拘', '不拘', '不拘', '不拘', '不拘', '不拘', '不拘', 'PowerPoint', 'HTML', 'Windows 2000', '不拘', '不拘', '不拘', '不拘', 'Excel', '不拘', 'Excel', '不拘', 'Excel', '不拘', '不拘', 'Excel', '不拘', 'Excel', '不拘', '不拘', '不拘', 'Adobe Photoshop', '不拘', '不拘', 'Windows 2000', '不拘', '不拘', '不拘', 'HTML', 'Excel', '不拘', '不拘', '不拘', '不拘', '不拘', '不拘', '不拘', 'Windows 2000', 'Excel', '不拘', '不拘', '不拘', '不拘', 'Excel', '不拘', '不拘', '不拘', '不拘', 'Excel', '不拘', '不拘', '不拘', 'PowerPoint', 'Excel', '不拘', '不拘', '不拘', '不拘', '不拘', '不拘', '不拘', 'PowerPoint', 'HTML', 'Windows 2000', '不拘', '不拘', '不拘', '不拘', 'Excel', '不拘', 'Excel', '不拘', 'Excel', '不拘', '不拘', 'Excel', '不拘', 'Windows 2000', 'Excel', '不拘', '不拘', '不拘', '不拘', 'Excel', '不拘', '不拘', '不拘', '不拘', 'Excel', '不拘', '不拘', '不拘', 'PowerPoint', 'Excel', '不拘', '不拘', '不拘', 'Windows 2000', 'Excel', '不拘', '不拘', '不拘', '不拘', 'Excel', '不拘', '不拘', '不拘', '不拘', 'Excel', '不拘

接著先觀察這個代表所有技能的alltools list裡面所有的獨特值：

In [6]:
set(alltools)

{'3ds Max',
 'Adobe Photoshop',
 'After Effects',
 'AutoCAD',
 'CoolDraw',
 'Excel',
 'Flash',
 'Google Analytics',
 'HTML',
 'Illustrator',
 None,
 'Outlook',
 'PhotoImpact',
 'PowerPoint',
 'Premiere',
 'Windows 2000',
 'Word',
 '不拘',
 '中文打字20~50',
 '中文打字50~75',
 '英文打字20~50',
 '英文打字50~75'}

再把這些獨特值裝成一個list：

In [7]:
list(set(alltools))

['3ds Max',
 None,
 'PowerPoint',
 'Adobe Photoshop',
 'Google Analytics',
 '中文打字50~75',
 '英文打字50~75',
 'Illustrator',
 'Outlook',
 'HTML',
 'Excel',
 'Premiere',
 'CoolDraw',
 'AutoCAD',
 'Windows 2000',
 'PhotoImpact',
 '不拘',
 '英文打字20~50',
 'Word',
 'After Effects',
 '中文打字20~50',
 'Flash']

用for迴圈確認所有技能的alltools裡面，每一個獨特值的資料筆數：

In [8]:
outlook = 0
Windows_2000 = 0
Excel = 0
Nonex = 0
英文打字2050 = 0
CoolDraw = 0
Premiere = 0
不拘 = 0
threedsMax = 0
AutoCAD = 0
Illustrator = 0
AdobePhotoshop = 0
HTML = 0
GoogleAnalytics = 0
英文打字5075 = 0 
PowerPoint = 0
中文打字5075 = 0
Flash = 0
AfterEffects = 0
Word = 0
中文打字2050 = 0
PhotoImpact = 0
list_to_sort = []
for i in alltools:
    if i == 'Outlook':
        outlook += 1
    elif i == 'Windows 2000':
        Windows_2000 += 1
    elif i == 'Excel':
        Excel += 1
    elif i == None:
        Nonex += 1
    elif i == '英文打字20~50':
        英文打字2050 += 1
    elif i == 'CoolDraw':
        CoolDraw += 1
    elif i == 'Premiere':
        Premiere += 1
    elif i == '不拘':
        不拘 += 1
    elif i == '3ds Max':
        threedsMax += 1
    elif i == 'AutoCAD':
        AutoCAD += 1
    elif i == 'Illustrator':
        Illustrator += 1
    elif i == 'Adobe Photoshop':
        AdobePhotoshop += 1
    elif i == 'HTML':
        HTML += 1
    elif i == 'Google Analytics':
        GoogleAnalytics += 1
    elif i == '英文打字50~75':
        英文打字5075 += 1
    elif i == 'PowerPoint':
        PowerPoint += 1
    elif i == '中文打字50~75':
        中文打字5075 += 1
    elif i == 'Flash':
        Flash += 1
    elif i == 'After Effects':
        AfterEffects += 1
    elif i == 'Word':
        Word += 1
    elif i == '中文打字20~50':
        中文打字2050 += 1
    elif i == 'PhotoImpact':
        PhotoImpact += 1

print('outlook:'+str(outlook), 
'Windows_2000:'+str(Windows_2000),
'Excel:'+str(Excel),
'Nonex:'+str(Nonex),
'英文打字2050:'+str(英文打字2050),
'CoolDraw:'+str(CoolDraw),
'Premiere:'+str(Premiere),
'不拘:'+str(不拘),
'3dsMax:'+str(threedsMax),
'AutoCAD:'+str(AutoCAD),
'Illustrator:'+str(Illustrator),
'AdobePhotoshop:'+str(AdobePhotoshop),
'HTML:'+str(HTML),
'GoogleAnalytics:'+str(GoogleAnalytics),
'英文打字5075:'+str(英文打字5075), 
'PowerPoint:'+str(PowerPoint),
'中文打字5075:'+str(中文打字5075),
'Flash:'+str(Flash),
'AfterEffects:'+str(AfterEffects),
'Word:'+str(Word),
'中文打字2050:'+str(中文打字2050),
'PhotoImpact:'+str(PhotoImpact))

outlook:151 Windows_2000:149 Excel:760 Nonex:20129 英文打字2050:149 CoolDraw:116 Premiere:149 不拘:2099 3dsMax:1 AutoCAD:1 Illustrator:266 AdobePhotoshop:418 HTML:70 GoogleAnalytics:79 英文打字5075:149 PowerPoint:877 中文打字5075:149 Flash:116 AfterEffects:116 Word:844 中文打字2050:186 PhotoImpact:116


我們就得到另一個結論：

<h3>行銷企劃人員前6大重要技能</h3>

1. PowerPoint: 877筆 <br>
2. Word: 844筆 <br>
3. Excel: 760筆 <br>
4. AdobePhot shop: 418筆 <br>
5. Illustrator: 266筆 <br>
6. 中文打字2050: 186筆 <br>

<h2>接下來找出至少英文要中等以上的資料筆數</h2>

用最簡單的方式直接計算：

In [9]:
data['lan'].value_counts()

不拘                                                               1861
 英文 -- 聽 /中等、說 /中等、讀 /中等、寫 /中等                                    517
 英文 -- 聽 /精通、說 /精通、讀 /精通、寫 /精通                                    226
 英文 -- 聽 /略懂、說 /略懂、讀 /略懂、寫 /略懂                                    150
 英文 -- 聽 /略懂、說 /略懂、讀 /略懂、寫 /略懂 日文 -- 聽 /略懂、說 /略懂、讀 /略懂、寫 /略懂      117
 英文 -- 聽 /精通、說 /精通、讀 /精通、寫 /精通 日文 -- 聽 /略懂、說 /略懂、讀 /略懂、寫 /略懂       70
 英文 -- 聽 /精通、說 /精通、讀 /精通、寫 /精通 台語 -- 精通                            33
 日文 -- 聽 /略懂、說 /略懂、讀 /略懂、寫 /略懂 英文 -- 聽 /精通、說 /精通、讀 /精通、寫 /精通       33
 英文 -- 聽 /中等、說 /中等、讀 /精通、寫 /精通 台語 -- 略懂                             1
 英文 -- 聽 /中等、說 /中等、讀 /精通、寫 /精通                                      1
 英文 -- 聽 /中等、說 /中等、讀 /中等、寫 /中等 中文 -- 聽 /精通、說 /精通、讀 /精通、寫 /精通        1
Name: lan, dtype: int64

用總筆數扣掉英文中等以下(也扣除不拘)的筆數：

In [10]:
english_above_ave = 3010-1861-150-117
english_above_ave

882

In [11]:
english_above_ave_v2 = 3010-150-117
english_above_ave_v2

2743

<h3>所以至少英文中等以上(扣除不拘)有882筆, 至少英文中等以上(保留不拘)有2743筆</h3>

<h3>接下來重新整理出一份新的csv檔案</h3>

只包含特定欄位，例如職稱、公司名稱等等

In [12]:
data.rename(columns={'title':'職稱',
                     'cName':'公司名稱',
                     'workT':'工作時間',
                     'hireNum':'最高需求人數(數字)',
                     'workExp':'要求年資（數字）'}, inplace=True)

In [13]:
cols = ['職稱','公司名稱','jobLoc','工作時間',
        '最高需求人數(數字)','salary','job_update','要求年資（數字）','edu']
data = data[cols]
data.head(3)

職稱           公司名稱               jobLoc            工作時間 最高需求人數(數字)  \
0  行銷企劃      華友旅行社有限公司    台北市大安區敦化南路二段76號8樓  日班，09:00~18:00        1 人   
1  行銷企劃  羊舍鮮羊乳_健豪農產品商行        彰化縣彰化市河濱路475號              日班        1 人   
2  行銷企劃     吉品養生股份有限公司  台北市大安區忠孝東路4段212號10樓              日班         不限   

              salary  job_update 要求年資（數字）       edu  
0  月薪 28,000~39,000元  2019-07-10     1年以上  專科、大學、碩士  
1  月薪 35,000~40,000元  2019-07-06     2年以上      大學以上  
2  月薪 30,000~40,000元  2019-06-20     2年以上      專科以上

開始切割公司地址變成縣市、鄉鎮、路名，加上這些新的欄位到主要資料後，再刪除原始欄位：

In [14]:
old_jobloc = list(data['jobLoc'])
city = []
town = []
road = []
for i in old_jobloc:
    city.append(i[0:3])
    town.append(i[3:6])
    road.append(i[6:])
data['city'] = city
data['town'] = town
data['road'] = road

data = data.drop(columns=['jobLoc'])
data.rename(columns={'city':'縣市',
                     'town':'鄉鎮',
                     'road':'路名'}, inplace=True)

data.head(3)

職稱           公司名稱            工作時間 最高需求人數(數字)             salary  \
0  行銷企劃      華友旅行社有限公司  日班，09:00~18:00        1 人  月薪 28,000~39,000元   
1  行銷企劃  羊舍鮮羊乳_健豪農產品商行              日班        1 人  月薪 35,000~40,000元   
2  行銷企劃     吉品養生股份有限公司              日班         不限  月薪 30,000~40,000元   

   job_update 要求年資（數字）       edu   縣市   鄉鎮             路名  
0  2019-07-10     1年以上  專科、大學、碩士  台北市  大安區    敦化南路二段76號8樓  
1  2019-07-06     2年以上      大學以上  彰化縣  彰化市        河濱路475號  
2  2019-06-20     2年以上      專科以上  台北市  大安區  忠孝東路4段212號10樓

接著整理出最高薪水：

In [15]:
old_salary = list(data['salary'])
new_salary = []
for i in old_salary:
    new_salary.append(i[10:16])
data['high_salary'] = new_salary

data = data.drop(columns=['salary'])
data.rename(columns={'high_salary':'最高月薪(數字)'},
                     inplace=True)

data.head(3)

職稱           公司名稱            工作時間 最高需求人數(數字)  job_update 要求年資（數字）  \
0  行銷企劃      華友旅行社有限公司  日班，09:00~18:00        1 人  2019-07-10     1年以上   
1  行銷企劃  羊舍鮮羊乳_健豪農產品商行              日班        1 人  2019-07-06     2年以上   
2  行銷企劃     吉品養生股份有限公司              日班         不限  2019-06-20     2年以上   

        edu   縣市   鄉鎮             路名 最高月薪(數字)  
0  專科、大學、碩士  台北市  大安區    敦化南路二段76號8樓   39,000  
1      大學以上  彰化縣  彰化市        河濱路475號   40,000  
2      專科以上  台北市  大安區  忠孝東路4段212號10樓   40,000

同樣整理工作職缺更新年、月份、日期：

In [16]:
old_job_update = list(data['job_update'])
year = []
month = []
date = []
for i in old_job_update:
    year.append(i[0:4])
    month.append(i[5:7])
    date.append(i[8:])
data['year'] = year
data['month'] = month
data['date'] = date

data = data.drop(columns=['job_update'])
data.rename(columns={'year':'更新年份',
                     'month':'更新月份',
                     'date':'更新日'},
                     inplace=True)

data.head(3)

職稱           公司名稱            工作時間 最高需求人數(數字) 要求年資（數字）       edu   縣市  \
0  行銷企劃      華友旅行社有限公司  日班，09:00~18:00        1 人     1年以上  專科、大學、碩士  台北市   
1  行銷企劃  羊舍鮮羊乳_健豪農產品商行              日班        1 人     2年以上      大學以上  彰化縣   
2  行銷企劃     吉品養生股份有限公司              日班         不限     2年以上      專科以上  台北市   

    鄉鎮             路名 最高月薪(數字)  更新年份 更新月份 更新日  
0  大安區    敦化南路二段76號8樓   39,000  2019   07  10  
1  彰化市        河濱路475號   40,000  2019   07  06  
2  大安區  忠孝東路4段212號10樓   40,000  2019   06  20

接著整理最低學歷，先看看學歷寫法：

In [17]:
data['edu'].drop_duplicates()

0       專科、大學、碩士
1           大學以上
2           專科以上
4             不拘
5          專科、大學
9             大學
11         大學、碩士
16      高中、專科、大學
41            碩士
45          碩士以上
2996        高中以上
Name: edu, dtype: object

看來抓前幾個字就是最低學歷要求：

In [18]:
old_edu = list(data['edu'])
edu = []
for i in old_edu:
    edu.append(i[0:2])
data['low_edu'] = edu

data = data.drop(columns=['edu'])
data.rename(columns={'low_edu':'最低學歷要求'},
                     inplace=True)

data.head(3)

職稱           公司名稱            工作時間 最高需求人數(數字) 要求年資（數字）   縣市   鄉鎮  \
0  行銷企劃      華友旅行社有限公司  日班，09:00~18:00        1 人     1年以上  台北市  大安區   
1  行銷企劃  羊舍鮮羊乳_健豪農產品商行              日班        1 人     2年以上  彰化縣  彰化市   
2  行銷企劃     吉品養生股份有限公司              日班         不限     2年以上  台北市  大安區   

              路名 最高月薪(數字)  更新年份 更新月份 更新日 最低學歷要求  
0    敦化南路二段76號8樓   39,000  2019   07  10     專科  
1        河濱路475號   40,000  2019   07  06     大學  
2  忠孝東路4段212號10樓   40,000  2019   06  20     專科

最後把全部更新過後的欄位按照自己想要的順序排列：

In [19]:
cols = ['職稱','公司名稱','縣市','鄉鎮','路名','工作時間',
        '最高需求人數(數字)','最高月薪(數字)','更新年份','更新月份','更新日','要求年資（數字）','最低學歷要求']
data = data[cols]
data.head(3)

職稱           公司名稱   縣市   鄉鎮             路名            工作時間 最高需求人數(數字)  \
0  行銷企劃      華友旅行社有限公司  台北市  大安區    敦化南路二段76號8樓  日班，09:00~18:00        1 人   
1  行銷企劃  羊舍鮮羊乳_健豪農產品商行  彰化縣  彰化市        河濱路475號              日班        1 人   
2  行銷企劃     吉品養生股份有限公司  台北市  大安區  忠孝東路4段212號10樓              日班         不限   

  最高月薪(數字)  更新年份 更新月份 更新日 要求年資（數字） 最低學歷要求  
0   39,000  2019   07  10     1年以上     專科  
1   40,000  2019   07  06     2年以上     大學  
2   40,000  2019   06  20     2年以上     專科

再看一次自己整理出來的新資料的全貌：

In [20]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3010 entries, 0 to 3009
Data columns (total 13 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   職稱          3010 non-null   object
 1   公司名稱        3010 non-null   object
 2   縣市          3010 non-null   object
 3   鄉鎮          3010 non-null   object
 4   路名          3010 non-null   object
 5   工作時間        3010 non-null   object
 6   最高需求人數(數字)  3010 non-null   object
 7   最高月薪(數字)    3010 non-null   object
 8   更新年份        3010 non-null   object
 9   更新月份        3010 non-null   object
 10  更新日         3010 non-null   object
 11  要求年資（數字）    3010 non-null   object
 12  最低學歷要求      3010 non-null   object
dtypes: object(13)
memory usage: 305.8+ KB


確認完畢就輸出成csv檔：

In [21]:
data.to_csv('pn110資料整理作業donev2.csv')

<h1>這次的任務就完成了！</h1>